core/shared/src/main/scala/zio/logging/internal/WriterProvider.scala

```diff
@@ -0,0 +1,73 @@
+package zio.logging.internal
+
+import java.io.{ BufferedWriter, FileOutputStream, OutputStreamWriter, Writer }
+import java.nio.charset.Charset
+import java.nio.file.{ FileSystems, Path }
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+private[logging] sealed trait WriterProvider {
+  def writer: Writer
+}
+
+private[logging] object WriterProvider {
+  final case class SimpleWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int]
+  ) extends WriterProvider {
+    override val writer: Writer = {
+      val output = new OutputStreamWriter(new FileOutputStream(destination.toFile, true), charset)
+      bufferedIOSize match {
+        case Some(bufferSize) => new BufferedWriter(output, bufferSize)
+        case None             => output
+      }
+    }
+  }
+
+  final case class TimeBasedRollingWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int],
+    formatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
+  ) extends WriterProvider {
+    import TimeBasedRollingWriterProvider._
+
+    private var timeInUse             = makeNewTime
+    private var currentWriter: Writer = makeWriter(makePath(destination, timeInUse, formatter), charset, bufferedIOSize)
+
+    override def writer: Writer = {
+      val newTime = makeNewTime
+      if (newTime != timeInUse) {
+        currentWriter.close()
+        currentWriter = makeWriter(makePath(destination, timeInUse, formatter), charset, bufferedIOSize)
+        timeInUse = newTime
```

Comment:
as i mentioned before, i think this implementation is not thread safe, some concurrency patterns should be use so only one caller will create new writer and close current one if it is needed

Reply:
Do I make it like the code below? Also, could you provide an example code for concurrent patterns?

```scala
    import java.util.concurrent.locks.ReentrantLock
    private val lock: ReentrantLock = new ReentrantLock()

    override def writer: Writer = {
      val newTime = makeNewTime
      if (newTime != timeInUse) {
        lock.lock()
        try {
          currentWriter.close()
          currentWriter = makeWriter(makePath(destination, timeInUse), charset, bufferedIOSize)
          timeInUse = newTime
        } finally {
          lock.unlock()
        }
      }
      currentWriter
    }
```

Reply:
this could work 

however i think time should be checked also under the lock

```scala
  override def writer: Writer = {
    val newTime = makeNewTime
    if (newTime != timeInUse) {
      lock.lock()
      try {
         if (newTime != timeInUse) {
            currentWriter.close() 
            currentWriter = makeWriter(makePath(destination, newTime), charset, bufferedIOSize) 
            timeInUse = newTime
        }
      } finally {
        lock.unlock()
      }
    } 
     currentWriter
  }
```
or just

```scala
 override def writer: Writer = {
    val newTime = makeNewTime
    
      lock.lock()
      try {
         if (newTime != timeInUse) {
            currentWriter.close() 
            currentWriter = makeWriter(makePath(destination, newTime), charset, bufferedIOSize) 
            timeInUse = newTime
        }
      } finally {
        lock.unlock()
      }
    
     currentWriter
  }
  ```
  
  i think that first version should be less locking, even if it is probably less correct